Add hook for elasticsearch query and option to remove query body

joeyfromspace opened the pull request:
### What does this PR do?
Adds two configuration options to the `elasticsearch` plugin:

1. A boolean called `includeBody` that when set to `false` will not include `elasticsearch.body` in shipped spans
2. Similar `hooks` implementation as found in the `http` plugin that allows an arbitrary function to operate on the span before it is closed

### Motivation
In a HIPAA-compliant environment, we need the ability to redact or at least exclude elements of traces which may contain PII. This PR will give us the tools to do that for `elasticsearch` queries.

### Plugin Checklist

- [X] Unit tests.
- [X] TypeScript [definitions][1].
- [X] TypeScript [tests][2].
- [X] API [documentation][3].
- [X] CircleCI [jobs/workflows][4].
- [X] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### Additional Notes
As this is my first contribution, please let me know if I failed to adhere to any particular patterns or if some of my decisions are not consistent with the vision of the library. Thanks!

rochdev commented:
What is the use case for the hook? Mostly asking because only the span is passed right now, and I wonder if it would be more useful to pass other things as well.

About the `elasticsearch.body` tag, It's also worth noting that our agent obfuscates the body automatically. So you would see the structure of the JSON but not the values.

joeyfromspace commented:
> What is the use case for the hook? Mostly asking because only the span is passed right now, and I wonder if it would be more useful to pass other things as well.
> 
> About the `elasticsearch.body` tag, It's also worth noting that our agent obfuscates the body automatically. So you would see the structure of the JSON but not the values.

Thanks for looking! We would use the hook to run a function that examines the body for PII (like email addresses and phone numbers) and automatically redacts them. We have a similar function in our application that does the same for URLs in the express and http hooks.

Maybe this is a matter for support, but the agent is not obfuscating the bodies for us. If we turn the plugin on, we can plainly see the values being queried in traces. The mongodb plugin however _does_ obfuscate and replaces query values with `?`.

rochdev commented:
> We would use the hook to run a function that examines the body for PII (like email addresses and phone numbers) and automatically redacts them.

Given the hook and the ability to redact, is it still useful to keep the `includeBody` option? I wonder if maybe an option to redact every value in the body instead could be more flexible. This is supposed to already happen automatically however.

For the hook, I would pass the params object as a second argument since this would allow for more use cases where you would need information about the query in the hook.

> Maybe this is a matter for support, but the agent is not obfuscating the bodies for us.

It has to be enabled in the agent config like this:

```yaml
apm_config:
  obfuscation:
    elasticsearch:
      enabled: true
```

If this is already the case and the body doesn't get obfuscated, definitely contact support as this should never happen.

joeyfromspace commented:
@rochdev Thanks! I've contacted support re: the obfuscation configuration.

I've also updated the code to include params, as that is a useful tool I think. I've also removed `includeBody` flag.

rochdev commented on index.d.ts:
```diff
@@ -642,7 +642,17 @@ declare namespace plugins {
+      /**
+       * Hook to execute just before the query span finishes.
+       */
+      query?: (span?: opentracing.Span, params?: any) => any;
```
The type of params should probably be `TransportRequestParams` from [Transport.d.ts](https://github.com/elastic/elasticsearch-js/blob/master/lib/Transport.d.ts).

joeyfromspace replied:
I pulled `@elastic/elasticsearch` in as a dev dependency so I could import it in this test. Though let me know if you prefer me to simply recreate the interface rather than add another package into the mix.

rochdev replied:
Ideally it would be best to use the types from the library in case they change over time. While it would be better to avoid having to add the library as a dev dependency, this is a limitation that will be addressed later for all plugins.

joeyfromspace replied:
Ah okay, that's fine. How should I get around the licensing restriction in the linter?

rochdev commented on packages/datadog-plugin-elasticsearch/src/index.js:
```diff
@@ -36,17 +37,17 @@ function createWrapRequest (tracer, config) {
 
           if (promise && typeof promise.then === 'function') {
-            promise.then(() => finish(span), e => finish(span, e))
+            promise.then(() => finish(span, null, params, config), e => finish(span, e, params, config))
```
Since the only optional argument is `e`, it should go last to avoid the `null`.

rochdev commented on packages/datadog-plugin-elasticsearch/test/index.spec.js:
```diff
@@ -215,7 +215,10 @@ describe('Plugin', () => {
-          return agent.load(plugin, 'elasticsearch', { service: 'test' })
+          return agent.load(plugin, 'elasticsearch', {
+            service: 'test',
+            hooks: { query: span => { span.setTag('elasticsearch.params', 'foo') } }
```
This doesn't test accessing `params`.

rochdev commented on LICENSE-3rdparty.csv:
```diff
@@ -29,6 +29,7 @@ require,url-parse,MIT,Copyright 2015 Unshift.io Arnout Kazemier the Contributors
 require,whatwg-fetch,MIT,Copyright 2014-2016 GitHub Inc.
 dev,@babel/core,MIT,Copyright 2014-present Sebastian McKenzie and other contributors
 dev,@babel/preset-env,MIT,Copyright 2014-present Sebastian McKenzie and other contributors
+dev,@elastic/elasticsearch,Apache license 2.0,Copyright 2013 Elasticsearch B.V.
```
I'm not 100% sure it matters, but just to be safe we usually write this license as `Apache-2.0`.

rochdev commented on index.d.ts:
```diff
@@ -1,6 +1,7 @@
 import { ClientRequest, IncomingMessage, ServerResponse } from "http";
 import * as opentracing from "opentracing";
 import { SpanOptions } from "opentracing/lib/tracer";
+import { TransportRequestParams } from "@elastic/elasticsearch/lib/Transport"
```
I just realized that this introduces a hard dependency on the library. I guess there is no other option than to copy the definition as you mentioned earlier. Sorry about the oversight.

joeyfromspace replied:
No problem! I just took out the package and followed the pattern for the `EventEmitter` interface when adding the interface type.
